index.bs

```diff
@@ -261,39 +259,78 @@ vrDisplay.requestPresent([{ source: canvas }]).then(function() {
+</pre>
+
+## CanvasVRLayer ## {#interface-canvasvrlayer}
+
+The {{CanvasVRLayer}} describes a layer containing head tracked stereo content
+rendered to a canvas with a WebGL context.
+
+<pre class="idl" id="canvasvrlayer-interface">
 typedef (HTMLCanvasElement or
-         OffscreenCanvas) VRSource;
+         OffscreenCanvas) CanvasVRSource;
 
-[Constructor(optional VRLayerInit layer)]
-interface VRLayer {
-  readonly attribute VRSource? source;
+[Constructor(optional CanvasVRSource source)]
+interface CanvasVRLayer : VRLayer {
+  attribute CanvasVRSource source;
 
-  readonly attribute sequence&lt;float&gt; leftBounds;
-  readonly attribute sequence&lt;float&gt; rightBounds;
-};
+  attribute float renderScale;
+  attribute boolean resizeSourceOnPresent;
+
+  readonly attribute unsigned long renderWidth;
+  readonly attribute unsigned long renderHeight;
 
-dictionary VRLayerInit {
-  VRSource? source = null;
+  void setLeftBounds(float left, float bottom, float right, float top);
+  sequence&lt;float&gt; getLeftBounds();
 
-  sequence&lt;float&gt; leftBounds = [ ];
-  sequence&lt;float&gt; rightBounds = [ ];
+  void setRightBounds(float left, float bottom, float right, float top);
+  sequence&lt;float&gt; getRightBounds();
 };
 </pre>
 
-### Attributes ### {#vrlayer-attributes}
+### Attributes ### {#canvasvrlayer-attributes}
+
+<dfn attribute for="CanvasVRLayer">source</dfn>
+The {{CanvasVRLayer/source}} attribute defines the canvas whose contents will be presented for this layer by the {{VRDisplay}} when {{VRDisplay}}.{{submitFrame()}} is called. Upon being passed into {{requestPresent()}} the current backbuffer of the source's context MAY be lost, even if the context was created with the <code>preserveDrawingBuffer</code> context creation attribute set to <code>true</code>. Calls to {{requestPresent()}} MUST be rejected if the {{CanvasVRLayer/source}} of any {{CanvasVRLayer}} provided does not represent a canvas with a <code>webgl</code> or <code>webgl2</code> context.
+
+<dfn attribute for="CanvasVRLayer">renderScale</dfn>
+Specifies the ratio of the number of canvas pixels to the {{VRDisplay}}'s physical pixels at the center of the output image. Note that a value of <code>1.0</code> will generally result in a larger canvas resolution than the {{VRDisplay}}'s physical screen in order to allow for corrective distiorion. Specifying a lower value can improve performance. A value of <code>0.0</code> instructs the user agent to select a ratio that provides a good balance between quality and performance for most applications on the current {{VRDisplay}}. Calls to {{requestPresent()}} MUST be rejected if the renderScale of any {{CanvasVRLayer}} provided is less than <code>0.0</code>.
+
+If this value was set to <code>0.0</code> when the {{CanvasVRLayer}} is passed to {{requestPresent()}} the {{renderScale}} of the corresponding {{CanvasVRLayer}} returned from {{getLayers()}} should be set to the value selected by the user agent.
+
```

Comment:
Good question. Demonstrates that the text needs clarification.

The intent is not to have the system doing dynamic scaling. The developer should be able to do that via viewport alteration, but the user agent won't attempt to guess at how the scaling should work. Instead the 0.0 case here indicates that the size reported should be a static value that gives the developer some wiggle room with the fill rate while still looking decent. Up to the UA what this means, and it's intended to be a value that we can tune fairly opaquely when needed. I would expect that in the real world most desktop devices will report a 1:1 value anyway, since they have the hardware to back it up, and mobile devices will report something lower. (GearVR docs recommend 1024x1024 per eye, for example, and Daydream sample code suggests using 7/10 the full 1:1 size)